### PR TITLE
Add multipart support

### LINKED ISSUE #8: Allow to upload files from disk

### PR DESCRIPTION
**Pull request by rossnomann**
Closes #8

### CODE REVIEW

File: tgbot/src/api.rs [hunk 1]
```diff
@@ -98,7 +99,7 @@ impl Api {
     }
 
     /// Executes a method
-    pub fn execute<M: Method>(&self, method: &M) -> ApiFuture<M::Response>
+    pub fn execute<M: Method>(&self, method: M) -> ApiFuture<M::Response>
```
Comment (ark0f) [inline tgbot/src/api.rs:102]: Return it back to `method: &M`

File: tgbot/src/methods/method.rs [hunk 2]
```diff
@@ -1,74 +1,11 @@
...
-    fn build(&self, base_url: &str, token: &str) -> String {
-        format!("{}/bot{}/{}", base_url, token, self.0)
-    }
-}
-
-#[derive(Clone, Debug)]
-pub(crate) enum RequestBody {
-    Json(Vec<u8>),
-    Empty,
+    fn into_request(self) -> Result<RequestBuilder, Error>;
```
Comment (ark0f) [inline tgbot/src/methods/method.rs:10]: And here

File: tgbot/src/methods/chat/set_photo.rs [hunk 3]
```diff
@@ -10,10 +13,9 @@ use serde::Serialize;
 ///
 /// Note: In regular groups (non-supergroups), this method will only work
 /// if the ‘All Members Are Admins’ setting is off in the target group
-#[derive(Clone, Debug, Serialize)]
+#[derive(Debug)]
```
Comment (ark0f) [inline tgbot/src/methods/chat/set_photo.rs:16]: Return `Clone` derives for structures

**Review by ark0f — CHANGES_REQUESTED** [review]

File: tgbot/src/api.rs (hunk 1 above)
Comment (mexus) [reply tgbot/src/api.rs:102]: `method: M` sounds better

File: tgbot/src/methods/chat/set_photo.rs (hunk 3 above)
Comment (mexus) [reply tgbot/src/methods/chat/set_photo.rs:16]: +1, might be useful, the rule of thumb is to add `Clone` to all public types

File: tgbot/src/methods/method.rs (hunk 2 above)
Comment (mexus) [reply tgbot/src/methods/method.rs:10]: it seems better this way (with `self` instead of `&self`)

File: tgbot/src/request/form.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+pub(crate) struct Form {
+    fields: HashMap<String, FormValue>,
+}
+
+impl Form {
+    pub(crate) fn new() -> Self {
+        Self { fields: HashMap::new() }
+    }
+
+    pub(crate) fn set_field<N, V>(&mut self, name: N, value: V)
```
Comment (mexus) [inline tgbot/src/request/form.rs]: mb `insert_field`?

File: tgbot/src/request/form.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    where
+        N: Into<String>,
+        V: Into<FormValue>,
+    {
+        self.fields.insert(name.into(), value.into());
+    }
+
+    pub(crate) fn into_multipart(self) -> MultipartForm<'static> {
```
Comment (mexus) [inline tgbot/src/request/form.rs]: Why not implement `From<Form> for MultipartForm<'static>`?

**Review by mexus — COMMENTED** [review]
:+1:

File: tgbot/src/methods/chat/set_photo.rs (hunk 3 above)
Comment (rossnomann) [reply tgbot/src/methods/chat/set_photo.rs:16]: You cant clone because of Box<Read + Send>

Comment (mexus) [reply tgbot/src/methods/chat/set_photo.rs:16]: ^ retracted

File: tgbot/src/request/form.rs (hunk 5 above)
Comment (rossnomann) [reply tgbot/src/request/form.rs]: done

File: tgbot/src/request/form.rs (hunk 4 above)
Comment (rossnomann) [reply tgbot/src/request/form.rs]: done

**Pull request merged**